tests/lms/views/decorators/test_h_api.py

```diff
@@ -0,0 +1,147 @@
+        util.generate_username.side_effect = MissingToolConsumerIntanceGUIDError()
+
+        with pytest.raises(HTTPBadRequest, match="tool_consumer_instance_guid"):
+            create_h_user(pyramid_request, mock.sentinel.jwt)
+
+    def test_it_400s_if_generate_username_raises_MissingUserIDError(self, create_h_user, pyramid_request, util):
+        util.generate_username.side_effect = MissingUserIDError()
+
+        with pytest.raises(HTTPBadRequest, match="user_id"):
+            create_h_user(pyramid_request, mock.sentinel.jwt)
+
+    def test_it_400s_if_generate_provider_raises_MissingToolConsumerInstanceGUIDError(self, create_h_user, pyramid_request, util):
+        util.generate_provider.side_effect = MissingToolConsumerIntanceGUIDError()
+
+        with pytest.raises(HTTPBadRequest, match="tool_consumer_instance_guid"):
+            create_h_user(pyramid_request, mock.sentinel.jwt)
+
+    def test_it_400s_if_generate_provider_unique_id_raises_MissingUserIDError(self, create_h_user, pyramid_request, util):
+        util.generate_provider_unique_id.side_effect = MissingUserIDError()
+
+        with pytest.raises(HTTPBadRequest, match="user_id"):
+            create_h_user(pyramid_request, mock.sentinel.jwt)
+
+    def test_it_creates_the_user_in_h(self, create_h_user, post, pyramid_request):
+        create_h_user(pyramid_request, mock.sentinel.jwt)
+
+        post.assert_called_once_with(
+            "https://example.com/api/users",
+            auth=("TEST_CLIENT_ID", "TEST_CLIENT_SECRET"),
+            data=json.dumps({
+                "username": "test_username",
+                "display_name": "test_display_name",
+                "authority": "TEST_AUTHORITY",
+                "identities": [{
+                    "provider": "test_provider",
+                    "provider_unique_id": "test_provider_unique_id",
+                }],
+            }),
+            timeout=1,
+        )
+
+    def test_it_504s_if_the_h_request_times_out(self, create_h_user, patch, post, pyramid_request):
+        post.side_effect = requests.exceptions.ReadTimeout()
+
+        with pytest.raises(HTTPGatewayTimeout):
+            create_h_user(pyramid_request, mock.sentinel.jwt)
+
+    def test_it_continues_to_the_wrapped_function_if_h_200s(self, create_h_user, pyramid_request, wrapped):
+        returned = create_h_user(pyramid_request, mock.sentinel.jwt)
+
+        wrapped.assert_called_once_with(pyramid_request, mock.sentinel.jwt)
+        assert returned == wrapped.return_value
+
+    def test_it_continues_to_the_wrapped_function_if_h_409s(self, create_h_user, post, pyramid_request, wrapped):
+        post.return_value.status_code = 409
+
+        returned = create_h_user(pyramid_request, mock.sentinel.jwt)
+
+        wrapped.assert_called_once_with(pyramid_request, mock.sentinel.jwt)
+        assert returned == wrapped.return_value
+
+    @pytest.mark.parametrize("status", (500, 501, 502, 503, 504, 400, 401, 403, 404, 408))
```

Comment:
This is really nice. It is explicit about what errors we want to cover with our 502 fallback...